pkg/pod/entrypoint.go

```diff
@@ -0,0 +1,174 @@
+			}
+		default:
+			// All other steps wait for previous file, write next file.
+			argsForEntrypoint = []string{
+				"-wait_file", filepath.Join(mountPoint, fmt.Sprintf("%d", i-1)),
+				"-post_file", filepath.Join(mountPoint, fmt.Sprintf("%d", i)),
+			}
+		}
+
+		cmd, args := s.Command, s.Args
+		if len(cmd) == 0 {
+			panic("oh no bad stuff") // TODO remove
+		}
+		if len(cmd) > 1 {
+			args = append(cmd[1:], args...)
+			cmd = []string{cmd[0]}
+		}
+		argsForEntrypoint = append(argsForEntrypoint, "-entrypoint", cmd[0], "--")
+		argsForEntrypoint = append(argsForEntrypoint, args...)
+
+		steps[i].Command = []string{entrypointBinary}
+		steps[i].Args = argsForEntrypoint
+		steps[i].VolumeMounts = append(steps[i].VolumeMounts, ToolsMount)
+	}
+	// Mount the Downward volume into the first step container.
+	steps[0].VolumeMounts = append(steps[0].VolumeMounts, DownwardMount)
+
+	return toolsInit, steps
+}
+
+// UpdateReady updates the Pod's annotations to signal the first step to start
+// by projecting the ready annotation via the Downward API.
+func UpdateReady(kubeclient kubernetes.Interface, pod corev1.Pod) error {
+	newPod, err := kubeclient.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error getting Pod %q when updating ready annotation: %w", pod.Name, err)
+	}
+
+	// Update the Pod's "READY" annotation to signal the first step to
+	// start.
+	if newPod.ObjectMeta.Annotations == nil {
+		newPod.ObjectMeta.Annotations = map[string]string{}
+	}
+	if newPod.ObjectMeta.Annotations[ReadyAnnotation] != ReadyAnnotationValue {
+		newPod.ObjectMeta.Annotations[ReadyAnnotation] = ReadyAnnotationValue
+		if _, err := kubeclient.CoreV1().Pods(newPod.Namespace).Update(newPod); err != nil {
+			return fmt.Errorf("Error adding ready annotation to Pod %q: %w", pod.Name, err)
+		}
+	}
+	return nil
+}
+
+// StopSidecars updates sidecar containers in the Pod to a nop image, which
+// exits successfully immediately.
+func StopSidecars(nopImage string, kubeclient kubernetes.Interface, pod corev1.Pod) error {
+	newPod, err := kubeclient.CoreV1().Pods(pod.Namespace).Get(pod.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error getting Pod %q when stopping sidecars: %w", pod.Name, err)
+	}
+
+	updated := false
+	if newPod.Status.Phase == corev1.PodRunning {
+		for _, s := range newPod.Status.ContainerStatuses {
+			if IsContainerSidecar(s.Name) && s.State.Running != nil {
+				for j, c := range newPod.Spec.Containers {
+					if c.Name == s.Name && c.Image != nopImage {
+						updated = true
+						newPod.Spec.Containers[j].Image = nopImage
+					}
+				}
+			}
+		}
+	}
+	if updated {
+		if _, err := kubeclient.CoreV1().Pods(newPod.Namespace).Update(newPod); err != nil {
+			return fmt.Errorf("Error adding ready annotation to Pod %q: %w", pod.Name, err)
+		}
+	}
+	return nil
+}
+
+func IsContainerStep(name string) bool    { return strings.HasPrefix(name, "step-") }
```

Comment:
v minor nit, maybe use the constants defined at the top of this file instead of hte string literals?  `StepPrefix`, `SidecarPrefix`

Reply:
Done!